haystack/core/pipeline/base.py

```diff
@@ -1042,24 +1042,134 @@ def _add_missing_input_defaults(component_inputs: Dict[str, Any], component_inpu
 
         return component_inputs
 
-    @staticmethod
+    def _tiebreak_waiting_components(
+        self,
+        component_name: str,
+        priority: ComponentPriority,
+        priority_queue: FIFOPriorityQueue,
+        topological_sort: Union[Dict[str, int], None],
+    ):
+        """
+        Decides which component to run when multiple components are waiting for inputs with the same priority.
+        """
+        components_with_same_priority = [component_name]
+
+        while len(priority_queue) > 0:
+            next_priority, next_component_name = priority_queue.peek()
+            if next_priority == priority:
+                priority_queue.pop()  # actually remove the component
+                components_with_same_priority.append(next_component_name)
+            else:
+                break
+
```

Comment:
I think this relies on the components in the priority queue to be grouped by priority type. Is that a good assumption to make? Unsure if we could end up with something like `priority_queue = [DEFER, DEFER_LAST, DEFER]`

Reply:
They are always sorted by priority, so we can never have a situation where we get [DEFER, DEFER_LAST, DEFER]. When you look at the implementation of `FIFOPriorityQueue` you can see that we sort by component priority first and then by insertion order into the queue (to have a stable sort for components with the same priority).